File: Graphs/P012.py
def dfs_traversal(n,edge):
    adj=[[] for _ in range(n)]
    for r,c in edge:
        adj[r].append(c)
        adj[c].append(r)
    visited=[False]*n
    res=[]
    def dfs(node):
        visited[node]=True
        res.append(node)
        for nei in adj[node]:
            if not visited[nei]:
                dfs(nei)
    for i in range(n):
        if not visited[i]:
            dfs(i)
    return res

n = 5
edges = [[0,1],[0,2],[1,3],[1,4]]

print(dfs_traversal(n, edges))

File: Striver_C++_Problems/Graphs/P012.cpp
// P012.cpp - DFS traversal of undirected graph

#include <bits/stdc++.h>
using namespace std;

vector<int> dfs_traversal(int n, const vector<pair<int,int>>& edges) {
    vector<vector<int>> adj(n);
    for (auto &e : edges) {
        int r = e.first, c = e.second;
        adj[r].push_back(c);
        adj[c].push_back(r);
    }
    vector<bool> visited(n, false);
    vector<int> res;

    function<void(int)> dfs = [&](int node) {
        visited[node] = true;
        res.push_back(node);
        for (int nei : adj[node]) {
            if (!visited[nei]) dfs(nei);
        }
    };

    for (int i = 0; i < n; i++) {
        if (!visited[i]) dfs(i);
    }
    return res;
}

int main() {
    int n = 5;
    vector<pair<int,int>> edges = {{0,1},{0,2},{1,3},{1,4}};

    auto ans = dfs_traversal(n, edges);
    for (int v : ans) cout << v << " ";
    cout << "\n";
    return 0;
}
